docs/advanced/UsageWithTypescript.md

```diff
@@ -56,47 +56,45 @@ Chat Action Constants & Shape:
 // src/store/chat/types.ts
 
 export enum ChatActions {
```

Comment:
I'm still not sure about the use of this enum.

One of the problems of using an enum here is that it is only possible to have type correctness if you refer to the enum. That is, the following is a compiler error:


```typescript
const foo: SendMessageAction = {
  type: 'SEND_MESSAGE', // "SEND_MESSAGE" is not assignable to `ChatActions`
  payload: ...
}
```

Which feels surprising. Is the motivation for using an enum just to prevent duplication of string literals?

Reply:
This is a valid concern, the motivation is definitely for preventing duplication of string literals. If one has to type the same string literal twice then there is a chance for an error/typo which typescript will consider valid (as far as I am concerned).

Regarding the issue you pointed out it does become a problem if you use actions with string literals. This doesn't become a concern if you strictly use action creators (assuming the action creators' type points to the enum).

Reply:
> there is a chance for an error/typo which typescript will consider valid (as far as I am concerned).

This is not true.

```ts
interface SendMessageAction {
  type: "SEND_MESSAGE",
  payload: ...
}

const foo: SendMessageAction {
  type: "foo", // compiler error: Type string is not assignable to "SEND_MESSAGE"
  payload: ...
}

const foo: SendMessageAction {
  type: "SEND_MESSAGE", // no problem
  payload: ...
}

function reducer(state: any, action: SendMessageAction | { type: unknown }) {
  switch (action.type) {
    case 'SEND_MESSAGE':
      // type is narrowed to SendMessageAction
      return state
  }
  return state
}
```